src/source/tile.js

```diff
@@ -527,6 +531,18 @@ class Tile {
             if (!sourceLayer || !sourceLayerStates || Object.keys(sourceLayerStates).length === 0) continue;
 
             bucket.update(sourceLayerStates, sourceLayer, availableImages, this.imageAtlas && this.imageAtlas.patternPositions || {});
+            if (painter._terrain && painter._terrain.enabled &&
+                (bucket instanceof LineBucket ||
+                bucket instanceof FillBucket ||
+                bucket instanceof FillExtrusionBucket ||
+                bucket instanceof CircleBucket)
+                // Symbol bucket has no programConfigurations. This has to be spelled out for Flow.
+            ) {
+                const sourceCache = painter.style._getSourceCache(bucket.layers[0].source);
+                if (sourceCache && painter._terrain && bucket.programConfigurations.needsUpload) {
+                    painter._terrain._clearRenderCacheForTile(sourceCache.id, this.tileID);
+                }
+            }
```

Comment:
he static analyzer can be finicky but I don't think we need to double check against terrain.
```suggestion
            bucket.update(sourceLayerStates, sourceLayer, availableImages, this.imageAtlas && this.imageAtlas.patternPositions || {});
            if (bucket instanceof LineBucket ||
                bucket instanceof FillBucket ||
                bucket instanceof FillExtrusionBucket ||
                bucket instanceof CircleBucket)
            {
                const sourceCache = painter.style._getSourceCache(bucket.layers[0].source);
                if (painter._terrain && painter._terrain.enabled && sourceCache && bucket.programConfigurations.needsUpload) {
                    painter._terrain._clearRenderCacheForTile(sourceCache.id, this.tileID);
                }
            }
```

Reply:
Good find @karimnaaji!